luigi_pipeline/lib/model/gcnv_mt_schema.py

```diff
@@ -0,0 +1,212 @@
+                        "LOF",
+                        "COPY_GAIN",
+                    )
+                },
+                {"gene_id": gene},
+            ),
+            self.mt.geneIds,
+        )
+
+    @row_annotation(fn_require=start)
+    def pos(self):
+        return self.mt.start
+
+    @row_annotation(fn_require=[contig, pos])
+    def xpos(self):
+        return variant_id.get_expr_for_xpos(
+            hl.locus(self.mt.contig, self.mt.pos)
+        )
+
+    @row_annotation(fn_require=xpos)
+    def xstart(self):
+        return self.mt.xpos
+
+    @row_annotation(fn_require=[contig, end])
+    def xstop(self):
+        return variant_id.get_expr_for_xpos(
+            hl.locus(self.mt.contig, self.mt.end)
+        )
+
+class SeqrGCNVGenotypesSchema(SeqrGenotypesSchema):
+
+    def __init__(self, *args, is_new_joint_call=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._is_new_joint_call = is_new_joint_call
+
+    @row_annotation(fn_require=SeqrGenotypesSchema.genotypes)
+    def samples(self):
+        return self._genotype_filter_samples(lambda g: True)
+
+    @row_annotation(fn_require=SeqrGenotypesSchema.genotypes)
+    def samples_new_call(self):
+        return self._genotype_filter_samples(lambda g: g.new_call)
+
+    def samples_no_call(self):
+        pass
+
+    def samples_num_alt(self):
+        pass
+
+    def samples_gq(self):
+        pass
+
+    def samples_ab(self):
+        pass
+
+    @row_annotation(fn_require=SeqrGenotypesSchema.genotypes)
+    def samples_qs(self, start=0, end=1000, step=10):
+        return hl.struct(**{
+            '%i_to_%i' % (i, i+step): self._genotype_filter_samples(lambda g: ((g.qs >= i) & (g.qs < i+step)))
+            for i in range(start, end, step)
+        }, **{
+            "samples_qs_gt_1000": self._genotype_filter_samples(lambda g: g.qs > 1000)
+        })
+
+    @row_annotation(name="samples_cn", fn_require=SeqrGenotypesSchema.genotypes)
+    def samples_cn(self, start=0, end=4, step=1):
+        return hl.struct(**{
+            '%i' % i: self._genotype_filter_samples(lambda g: g.cn == i)
+            for i in range(start, end, step)
+        }, **{
+            "samples_cn_gte_4": self._genotype_filter_samples(lambda g: g.cn >= 4)
+        })
+
+    def _genotype_filter_samples(self, filter):
+        samples = self.mt.genotypes.filter(filter).map(lambda g: g.sample_id)
+        return hl.if_else(hl.len(samples) > 0, samples, hl.missing(hl.dtype('array<str>')))
+    
+    def _genotype_fields(self):
+        return {
+            'sample_id': self.mt.sample_fix.first_match_in(SAMPLE_ID_REGEX)[0],
+            'qs': self.mt.QS,
+            'cn': self.mt.CN,
+            'defragged': self.mt.defragmented,
+            # Hail expression is to bool-ify a string value.
+            'prev_call': hl.if_else(hl.len(self.mt.identical_ovl) > 0, True, False) if self._is_new_joint_call else ~self.mt.is_latest,
```

Comment:
This was the gnarliest logic to reproduce (or at least caused me the most headaches!). Here's the old code for reference: https://github.com/broadinstitute/seqr-loading-pipelines/blob/master/sv_pipeline/exome/load_data.py#L121-L133

Reply:
You don't need to do `hl.if_else(condition, True, False)`, you can just do `condition`.  Also, the string "FALSE" is truthy so I don't think `~self.mt.is_latest` will work the way we want here

Reply:
Fixed the `if_else` bit!

`is_latest` has been getting type-handled by hail correctly ("TRUE" => `True` and "FALSE" => `False`) however, so at this point `mt.is_latest` is mapped appropriately.  it might be a bad assumption to let hail do this for us though... I'm not sure!

Reply:
Oh I didn't realize hail was imputing the type as a boolean, although that makes sense. We may want to either add some type checking on the matrix table, or actually add some specific typing when we import the table. My concern would be if a new value got added to the input data, we would not throw any error we would just have some undefined behavior. As I think about it, it probably is best to have explicit type definitions or the columns so the import fails if the data does not conform to what we expect